libfreerdp/core/license.c

```diff
@@ -132,21 +139,178 @@ static void license_print_product_info(LICENSE_PRODUCT_INFO* productInfo)
+		goto out;
+
+	for (i = 0; i < sizeof(hash); i++, hashStr += 2)
+		sprintf_s(hashStr, 3, "%.2x", hash[i]);
+
+    ret = TRUE;
+out:
+	winpr_Digest_Free(sha1);
+	return ret;
+}
+
+static BOOL saveCal(rdpSettings *settings, const BYTE *data, int length, char *hostname)
+{
+	char hash[41];
+	FILE *fp;
+	char *licenseStorePath = NULL;
+	char filename[MAX_PATH], filenameNew[MAX_PATH];
+	char *filepath, *filepathNew;
+	size_t written;
+	BOOL ret = FALSE;
+
+	if (!PathFileExistsA(settings->ConfigPath))
+	{
+		if (!PathMakePathA(settings->ConfigPath, 0))
+		{
+			WLog_ERR(TAG, "error creating directory '%s'", settings->ConfigPath);
+			goto fail;
+		}
+		WLog_INFO(TAG, "creating directory %s", settings->ConfigPath);
+	}
+
+	if (!(licenseStorePath = GetCombinedPath(settings->ConfigPath, licenseStore)))
+		goto fail;
+
+	if (!PathFileExistsA(licenseStorePath))
+	{
+		if (!PathMakePathA(licenseStorePath, 0))
+		{
+			WLog_ERR(TAG, "error creating directory '%s'", licenseStorePath);
+			goto fail_create_store_path;
+		}
+		WLog_INFO(TAG, "creating directory %s", licenseStorePath);
+	}
+
+	if (!computeCalHash(hostname, hash))
+		goto fail_compute_hash;
+	sprintf_s(filename, sizeof(filename)-1, "%s.cal", hash);
+	sprintf_s(filenameNew, sizeof(filenameNew)-1, "%s.cal.new", hash);
+
+	if (!(filepath = GetCombinedPath(licenseStorePath, filename)))
+		goto fail_file_path;
+
+	if (!(filepathNew = GetCombinedPath(licenseStorePath, filenameNew)))
+		goto fail_file_path_new;
+
+	fp = fopen(filepathNew, "wb");
+	if (!fp)
+		goto fail_file_open;
+
+	written = fwrite(data, length, 1, fp);
+	fclose(fp);
+
+	if (written != 1)
+	{
+		DeleteFile(filepathNew);
+		goto fail_file_open;
+	}
+
+	ret = MoveFileEx(filepathNew, filepath, MOVEFILE_REPLACE_EXISTING);
+
+fail_file_open:
+	free(filepathNew);
+fail_file_path_new:
+	free(filepath);
+fail_create_store_path:
+fail_compute_hash:
+fail_file_path:
+	free(licenseStorePath);
+fail:
+	return ret;
+}
+
+static BYTE *loadCalFile(rdpSettings *settings, const char* hostname, int *dataLen)
+{
+	char *licenseStorePath, *calPath;
+	char calFilename[MAX_PATH];
+	char hash[41];
+	int length, status;
+	FILE *fp;
+	BYTE *ret = NULL;
+
+	if (!computeCalHash(hostname, hash))
+	{
+		WLog_ERR(TAG, "loadCalFile: unable to compute hostname hash");
+		return NULL;
+	}
+
+	sprintf_s(calFilename, sizeof(calFilename)-1, "%s.cal", hash);
+
+	if (!(licenseStorePath = GetCombinedPath(settings->ConfigPath, licenseStore)))
+		goto fail;
+
+	if (!(calPath = GetCombinedPath(licenseStorePath, calFilename)))
+		goto error_calPath;
+
+	fp = fopen(calPath, "rb");
+	if (!fp)
+		goto error_open;
+
+	_fseeki64(fp, 0, SEEK_END);
+	length = _ftelli64(fp);
+	_fseeki64(fp, 0, SEEK_SET);
+
+	ret = (BYTE *)malloc(length);
+	if (!ret)
+		goto error_malloc;
+
+	status = fread(ret, length, 1, fp);
+	if (status <= 0)
+		goto error_read;
+
+	*dataLen = length;
+
+	fclose(fp);
+	free(calPath);
+	free(licenseStorePath);
+	return ret;
+
+error_read:
```

Comment:
initializing variables with `NULL` can avoid having to use multiple `goto` labels

Reply:
here I can simplify a little, but not completely getting rid of the 2 paths